### DISCUSSION TITLE
calling end_session_endpoint directly

### DISCUSSION THREAD

**Discussion by wzrdtales** [panva/node-oidc-provider#1279] [Ideas]
According to the readme this is supported, but as we discovered, actually only executing backchannel logout is supported, but not processing.

If a backchannel logout is performed against /session/end, currently it will just respond with the logout confirmation screen. So backchannel logout will effectively fail due to this. Keycloak handles the logout screen in this way, that it shows the screen if no id_token_hint is provided and automatically logs out if id_token_hint is present.

**Comment by panva** [comment]
Hello @wzrdtales 

> only executing backchannel logout is supported

That's the full extent of backchannel logout on an authorization server. I don't get where the "only" comes from.

Are you suggesting server-to-server call to the end_session_endpoint? That's not part of the backchannel logout spec and proprietary extensions to standard endpoints are not in scope.

**Reply by wzrdtales** [reply]
Actually this is well supported in all implementations of OIDC. In keycloak you can even trigger on the IDP configuration side in advanced "backchannel logout supported" to execute the logout action from the backend instead. I also did not find it in the official oidc spec, but inofficially it is already a standard that has been set basically through the community and I don't know many providers not supporting it.

**Reply by wzrdtales** [reply]
According to the keycloak docs it is part of the standard

> /realms/{realm-name}/protocol/openid-connect/logout/backchannel-logout
>
 >   This is the URL endpoint for performing backchannel logouts described in the OIDC specification.

But I couldn't find an actual reference for it though

**Reply by wzrdtales** [reply]
see this discussion also https://github.com/keycloak/keycloak/discussions/13707

**Reply by panva** [reply]
I am not aware of inofficial standards either. The end_session_endpoint is for user-involved logout as per the RP-Initiated Logout spec which defines it and adding proprietary behaviours to it is out of scope.

**Reply by wzrdtales** [reply]
This belongs to the rp initiated logout


> At the Logout Endpoint, the OP SHOULD ask the End-User whether to log out of the OP as well. Furthermore, the OP MUST ask the End-User this question if an id_token_hint was not provided or if the supplied ID Token does not belong to the current OP session with the RP and/or currently logged in End-User. If the End-User says "yes", then the OP MUST log out the End-User. 

The standard actually says exactly that, if the id_token_hint is present and valid, the consent screen is optional. This is the key difference in this module right now @panva . It can't be made optional yet

**Reply by wzrdtales** [reply]
so according to https://openid.net/specs/openid-connect-rpinitiated-1_0.html this is supported by the standard and not proprietary

**Reply by panva** [reply]
You're describing the IdP being a client in the Backchannel initiated logout spec. That has nothing to do with the end_session_endpoint. And it's also out of scope for this module because "how a user authenticates" and that entire lifecycle (whether its with an upstream IdP, credentials, or anything else) is up to you.

As such if you do establish end-user sessions in oidc-provider by federating upstream and want to expose a backchannel logout endpoint for those IdPs, that's up to you.

**Reply by wzrdtales** [reply]
Well no, the rp initiated logout spec says it.

> At the Logout Endpoint, the OP **SHOULD** ask the End-User whether to log out of the OP as well. Furthermore, the OP MUST ask the End-User this question if an id_token_hint was not provided or if the supplied ID Token does not belong to the current OP session with the RP and/or currently logged in End-User. If the End-User says "yes", then the OP MUST log out the End-User.

Important is the first sentence, the consent screen is optional. If the consent screen during /session/end could be skipped, the implementation would be according to spec and supporting the use case already supported everywhere else (okta, keycloak, ...)

**Reply by panva** [reply]
Apologies but you're mixing up things.

> If a backchannel logout is performed against /session/end

There is no such thing as **backchannel** logout against /session/end

You can make a user-agent (browser) POST/GET to it, but those are not "backchannel" that's either HTTP Form POST or HTTP Redirect.

> It can't be made optional yet

The fact that there is a prompt is up to you to configure behaviour for. You absolutely can come up with a configuration that auto-submits the confirmation prompt, making it effectively no-op and resolving the RP-Initiated Logout straight away.

But there isn't a way to have the /session/end endpoint do anything for requests made from outside the browser which is what backchannel would actually be, so - what is it that you're after?

**Reply by wzrdtales** [reply]
> There is no such thing as backchannel logout against /session/end

Sorry, don't blame me. Blame keycloak, I was researching myself for the spec defining backchannel logout against the OP, since keycloak named it completely wrong. Actually what we're talking about is just the RP initiated logout, which currently enforces the consent screen at all times, even though the standard allows otherwise if the id_token_hint is present and valid

**Reply by wzrdtales** [reply]
> But there isn't a way to have the /session/end endpoint do anything for requests made from outside the browser which is what backchannel would actually be, so - what is it that you're after?

yes, having the backend do the logout request for the user instead of going through the browser

**Reply by wzrdtales** [reply]
> The fact that there is a prompt is up to you to configure behaviour for. You absolutely can come up with a configuration that auto-submits the confirmation prompt, making it effectively no-op and resolving the RP-Initiated Logout straight away.

parsing html forms to submit them is an anti pattern and existing clients that support this behavior wont do that as well, so not the ideal solution. Being able to make the consent screen optional and directly confirm it instead when the id_token_hint is present would be the better option. I am looking if I can override the behavior of the lib right now in that regard, however I do think this should be a core behavior of the library itself.

**Reply by panva** [reply]
There's no such thing in the RP-Initiated Logout spec.

> An RP requests that the OP log out the End-User **by redirecting the End-User's User Agent** to the OP's Logout Endpoint.

> OpenID Providers MUST support the use of the HTTP GET and POST methods defined

None of this means you can throw a GET/POST request (neither from a server, nor a CORS one) at the endpoint. This implementation will handle such request as if there was no user session (because, well, there isn't one for the request you'd be making).

And so you can make it so that there's no end-user prompt through the `logoutSource` configuration helper. But it won't make the endpoint process a logout because in the request you'd be making with a GET/POST that isn't a full-page user-agent redirect there isn't any session/user involved.

**Reply by wzrdtales** [reply]
> There's no such thing in the RP-Initiated Logout spec.

There is, here it is: https://openid.net/specs/openid-connect-rpinitiated-1_0.html

> And so you can make it so that there's no end-user prompt through the logoutSource configuration helper. But it won't make the endpoint process a logout because in the request you'd be making with a GET/POST that isn't a full-page user-agent redirect there isn't any session/user involved.

The logout source can't execute a location 302/303 redirect right?

**Reply by panva** [reply]
> The logout source can't execute a location 302/303 redirect right?

no, but i can return HTML with the form hidden and self-submitted through javascript.

> There is, here it is: https://openid.net/specs/openid-connect-rpinitiated-1_0.html

The specification for RP-Initiated Logout does indeed allow there to be no prompt, which you can effectively achieve. But it does not mean you can GET/POST to it from a server. Only redirects from the user-agent (either 30x or full page html form post ones) are processed with an actual logout.

**Reply by wzrdtales** [reply]
The discussion appears to be very rigid, even though the standard clearly supports it. It is up to you however to decide that the library is the odd one out of all implementations. HTML + JS is no option

**Reply by panva** [reply]
> even though the standard clearly supports it

Supports what exactly? Omitting the prompt? Yeah, and you can achieve that. Using the endpoint through anything but user-agent redirects (either 30x or full page html form post ones)? No, the specification does not support such usage pattern.

**Reply by wzrdtales** [reply]
> Omitting the prompt? Yeah, and you can achieve that. 

How without copying the whole logic for /session/end and overriding it with a middleware?

**Reply by panva** [reply]
By rendering a self-submitting form instead.

```js
      logoutSource(ctx, form) {
        ctx.body = `<!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <title>Logout Request</title>
        <script nonce="${ctx.res.locals.cspNonce}">document.addEventListener('DOMContentLoaded', function () { document.forms[0].submit() })</script>
      </head>
      <body>
        <div style="display: none;">
          ${form}
          <input type="hidden" name="logout" form="op.logoutForm" value="yes"/>
        </div>
      </body>
      </html>`;
      },
    },
```


https://github.com/user-attachments/assets/676a1a43-4125-491f-b487-ebe66d62dd62

**Reply by wzrdtales** [reply]
which is not an option for real world use cases, which are backend based logout requests. Lots of software already implement those (just to name at least two nextcloud, keycloak, ...) .

But as I said before, your call, it is your library. I would be willing to drop the PR to make the consent optional (by configuration) with a valid token, but this discussion doesn't suggest you would be willing to merge it.

**Reply by panva** [reply]
It is a working option for the intended, spec-prescribed, and certified usage pattern. I did try and explain before that your intended usage pattern is off-spec and therefore out of scope. The `end_session_endpoint` requires the user-agent to be involved so that the OP can clear the browser session state.

**Reply by wzrdtales** [reply]
I see, so the issue is here in the implementation of the session handling itself, which relies solely on cookies and has no way of invalidating a session. An important realization, as this is problematic itself. I will see how we can strip out the session handling from the OIDC handling.

**Reply by wzrdtales** [reply]
having looked deeper at the code, I don't see a reason why the user-agent is required at all.

https://github.com/panva/node-oidc-provider/blob/abdcd950465650bc428ba60fcfc1db236ff4e43b/lib/models/interaction.js#L11

The accountId is effectively retrieved from the persistence adapter, and without the accountId and data from the persistence adapter the session should be effectively destroyed

**Reply by wzrdtales** [reply]
A quick test, deleting the key on the persistence layer, does indeed exactly what it should. So why do you think the user-agent needs to be involved? It does not.

**Reply by panva** [reply]
Just in case it wasn't clear, which I believe it was, I will not be pursuing support for off-spec behaviours. If you want to build a proprietary behaviour endpoint that intakes an artifact and removes other stored models that's fine, but it's just that - proprietary, and resides entirely in your deployment.

> So why do you think the user-agent needs to be involved? It does not.

> The `end_session_endpoint` requires the user-agent to be involved so that the OP can clear the browser session state.

That was a general statement for why the specification is the way it is, not necessarily aimed at this particular implementation.

**Reply by wzrdtales** [reply]
This is fine, we hard forked already. I just had the hope to make your library even better. Thank you for your work.

**Reply by panva** [reply]
Thank you for the kind words.
